SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,182 @@
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(USpatialPingComponent, ReplicatedPingID);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	// Attempt to cast the owner of this component to the PlayerController class.
+	// Component will do nothing if cast fails.
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (bStartWithPingEnabled)
+	{
+		SetPingEnabled(true);
+	}
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	Super::EndPlay(EndPlayReason);
+	// Clear up the timer before the component is disabled or destroyed.
+	SetPingEnabled(false);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	// Only execute if this component is attached to a player controller.
+	if (OwningController != nullptr)
+	{
+		// Only execute on owning local client.
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled)
+			{
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled)
+			{
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RoundTripPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		LastSentPingID = 0;
+		TimeoutCount = 0;
+		// Set looping timer to 'tick' this component with frequency matching minimum ping interval.
+		World->GetTimerManager().SetTimer(PingTickHandle, this, &USpatialPingComponent::TickPingComponent, MinPingInterval, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		bIsPingEnabled = false;
+		// Clear the timers.
+		World->GetTimerManager().ClearTimer(PingTickHandle);
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		// Reset ping output value.
+		RoundTripPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	SendNewPing();
+	// Pass latest measured ping to owning controller to be processed by PlayerState.
+	OwningController->UpdatePing(RoundTripPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	// Generate a new ping ID.
+	uint16 NewPingID = LastSentPingID + 1;
+	// Send and record new ping ID.
+	SendServerWorkerPingID(NewPingID);
+	LastSentPingID = NewPingID;
+	LastSentPingTimestamp = FPlatformTime::Seconds();
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		// Set a timeout timer to await a reply.
+		World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::OnPingTimeout, TimeoutLimit, false);
+	}
+}
+
+void USpatialPingComponent::OnPingTimeout()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		// Clear the timeout timer.
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
```

Comment:
This is a one-shot timer, so we don't need to clear it after it's run.

Reply:
Removed.